source/slang/ir-union.cpp

```diff
@@ -0,0 +1,776 @@
+            SLANG_ASSERT(structTypeLayout);
+
+            // We are going to emit code to extract each of the fields
+            // and collect them to use as operands to a `makeStruct`.
+            //
+            List<IRInst*> fieldVals;
+
+            // We need to walk over the fields in the order the IR expects them
+            UInt fieldCounter = 0;
+            for( auto irField : irStructType->getFields() )
+            {
+                IRType* fieldType = irField->getFieldType();
+
+                // TODO: We need to confirm/enforce that the fields of the
+                // IR struct and the fields of the layout still align.
+                //
+                UInt fieldIndex = fieldCounter++;
+                auto fieldLayout = structTypeLayout->fields[fieldIndex];
+                auto fieldTypeLayout = fieldLayout->getTypeLayout();
+
+                // The offset of the field can be computed from the base
+                // offset passed in, plus the reflection data for the field.
+                //
+                UInt64 fieldOffset = payloadOffset;
+                if(auto resInfo = fieldLayout->FindResourceInfo(LayoutResourceKind::Uniform))
+                    fieldOffset += resInfo->index;
+
+                // We make a recursive call to extract each field, expecting
+                // that this will bottom out eventually.
+                //
+                IRInst* fieldVal = extractPayload(
+                    taggedUnionInfo,
+                    taggedUnionVal,
+                    fieldType,
+                    fieldTypeLayout,
+                    fieldOffset);
+                fieldVals.Add(fieldVal);
+            }
+
+            // The final value is then just a new struct constructed from
+            // the extracted field values.
+            //
+            auto payloadVal = builder->emitMakeStruct(irStructType, fieldVals);
+            return payloadVal;
+        }
+        else if( auto vecType = as<IRVectorType>(payloadType) )
+        {
+            auto elementType = vecType->getElementType();
+
+            // We expect that by the time we are desugaring union types
+            // all vector types have literal constant values for their
+            // element count.
+            //
+            auto elementCountVal = vecType->getElementCount();
+            auto elementCountConst = as<IRIntLit>(elementCountVal);
+            SLANG_ASSERT(elementCountConst);
+            UInt elementCount = UInt(elementCountConst->getValue());
+
+            // HACK: There is currently no `VectorTypeLayout` and thus
+            // no way to query the layout of the elements of a vector
+            // type. Until that gets added we will kludge things here.
+            //
+            TypeLayout* elementTypeLayout = nullptr;
+            size_t elementSize = 0;
+            if(auto resInfo = payloadTypeLayout->FindResourceInfo(LayoutResourceKind::Uniform))
```

Comment:
I believe this code should be fine because it is only applying to vectors (not arrays in general) and the element type of a vector is always a basic type, and all of our targets seem to agree that for basic types size and alignment are the same, so that `elementSize` and `elementSize` should always be equal.

The best long-term fix will be to have a `VectorTypeLayout` that includes the layout for the element type, at which point these concerns are moot.